### PR TITLE
Added local registry support for baremetal

### PR DESCRIPTION
**Pull request by panktishah26**
*Issue #, if available:*

*Description of changes:*
Support for local registry for Baremetal provider

*Testing (if applicable):*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: pkg/executables/config/kind.yaml [hunk 1]
```diff
@@ -30,27 +30,28 @@ containerdConfigPatches:
...
+{{- if (ne .RegistryCACertPath "") }}
     - containerPath: /etc/containerd/certs.d
       hostPath: {{.RegistryCACertPath}}
       readOnly: true
 {{- end }}
-{{- end }}
 {{- if .DockerExtraMounts }}
-nodes:
-- role: control-plane
-  extraMounts:
```
[the remark this answers is not in the document]
Comment (panktishah26) [reply pkg/executables/config/kind.yaml:44]: For baremetal, we are using extraPortMapping which adds 

> nodes:
> - role: control-plane

in the kind config and on top of it if we add registrymirror it also adds the same nodes and role in the kind config which gives an error during kind cluster creation. 
https://github.com/aws/eks-anywhere/blob/main/pkg/executables/config/kind.yaml#L49

File: pkg/providers/tinkerbell/config/template-cp.yaml [hunk 2]
```diff
@@ -100,6 +100,14 @@ spec:
       bottlerocketBootstrap:
         imageRepository: {{.bottlerocketBootstrapRepository}}
         imageTag: {{.bottlerocketBootstrapVersion}}
+{{- end }}
+{{- if and .registryMirrorConfiguration (eq .format "bottlerocket") }}
```
[the remark this answers is not in the document]
Comment (panktishah26) [reply pkg/providers/tinkerbell/config/template-cp.yaml:104]: yes, I have tested it with BR and ubuntu both

File: pkg/executables/config/kind.yaml [hunk 3]
```diff
@@ -30,27 +30,28 @@ containerdConfigPatches:
         insecure_skip_verify = true
 {{- else }}
         ca_file = "/etc/containerd/certs.d/{{.RegistryMirrorEndpoint}}/ca.crt"
+{{- end }}
```
[the remark this answers is not in the document]
Comment (panktishah26) [reply pkg/executables/config/kind.yaml:33]: I guess you would be able to get it from this. It's difficult to get it from the diff. 
https://github.com/aws/eks-anywhere/blob/c2f7fbc0a35e5d4fd9a3e6820fe64bce30a604a8/pkg/executables/config/kind.yaml#L34

File: test/e2e/proxy_test.go [hunk 4]
```diff
@@ -71,3 +71,29 @@ func TestSnowKubernetes121UbuntuProxyConfig(t *testing.T) {
...
+		t,
+		framework.NewVSphere(t, framework.WithUbuntu122(),
+			framework.WithPrivateNetwork()),
+		framework.WithClusterFiller(api.WithControlPlaneCount(1)),
+		framework.WithClusterFiller(api.WithWorkerNodeCount(1)),
+		framework.WithClusterFiller(api.WithKubernetesVersion(v1alpha1.Kube122)),
+		framework.WithProxy(framework.VsphereProxyRequiredEnvVars),
+	)
+	runProxyConfigFlow(test)
+}
```
Comment (abhinavmpandey08) [inline test/e2e/proxy_test.go:86]: Why did you add this test? It looks like this is going to create a vSphere cluster with proxy config?

Comment (ptrivedi) [reply test/e2e/proxy_test.go:86]: Possibly to debug cli-tools container helm issue with proxy, lol

File: pkg/providers/tinkerbell/stack/stack.go [hunk 5]
```diff
@@ -311,3 +320,11 @@ func (s *Installer) CleanupLocalBoots(ctx context.Context, remove bool) error {
 	// finally, return an "already exists" error if boots exists and forceCleanup is not set
 	return errors.New("boots container already exists, delete the container manually or re-run the command with --force-cleanup")
 }
+
+func (s *Installer) url(originalURL string) string {
```
Comment (chrisdoherty4) [inline pkg/providers/tinkerbell/stack/stack.go:324]: Could we give this function a name that's a little more representative of what it does.

File: pkg/dependencies/factory.go [hunk 6]
```diff
@@ -282,7 +282,7 @@ func (f *Factory) WithProvider(clusterConfigFile string, clusterConfig *v1alpha1
 	case v1alpha1.DockerDatacenterKind:
 		f.WithDocker().WithKubectl()
 	case v1alpha1.TinkerbellDatacenterKind:
-		f.WithDocker().WithKubectl().WithWriter().WithHelm()
+		f.WithDocker().WithKubectl().WithWriter().WithHelm(executables.WithInsecure())
```
[the remark this answers is not in the document]
Comment (chrisdoherty4) [reply pkg/dependencies/factory.go:285]: I don't think we should assume to know what the user wants and should instead let them opt-in with config.

File: pkg/providers/tinkerbell/config/template-md.yaml [hunk 7]
```diff
@@ -86,6 +94,35 @@ spec:
             anonymous-auth: "false"
 {{- if .kubeletExtraArgs }}
 {{ .kubeletExtraArgs.ToYaml | indent 12 }}
+{{- end }}
+{{- if and .registryMirrorConfiguration (ne .format "bottlerocket") }}
+      files:
+{{- end }}
```
Comment (chrisdoherty4) [inline pkg/providers/tinkerbell/config/template-md.yaml:100]: I think this `end` is in the wrong place. It should wrap everything to be specified under `files` else there's a risk a susbequent condition could evaluate as true and populate invalid YAML (a YAML missing `files:`).

File: pkg/providers/tinkerbell/stack/stack.go [hunk 8]
```diff
@@ -102,13 +105,13 @@ func WithLoadBalancer() InstallOption {
 }
 
 // NewInstaller returns a Tinkerbell StackInstaller which can be used to install or uninstall the Tinkerbell stack
-func NewInstaller(docker Docker, filewriter filewriter.FileWriter, helm Helm, namespace, podCidrRange string) StackInstaller {
+func NewInstaller(docker Docker, filewriter filewriter.FileWriter, helm Helm, namespace string, clusterConfig *v1alpha1.Cluster) StackInstaller {
```
Comment (chrisdoherty4) [inline pkg/providers/tinkerbell/stack/stack.go:108]: Do we really need a complete cluster configuration to achieve stack installation? Adding the cluster object makes it ambiguous as to what is actually needed and how it impacts the behavior of the stack installer.

File: pkg/providers/tinkerbell/template.go [hunk 9]
```diff
@@ -402,10 +406,23 @@ func buildTemplateMapCP(clusterSpec *cluster.Spec, controlPlaneMachineSpec, etcd
...
 		"hardwareSelector":             controlPlaneMachineSpec.HardwareSelector,
 		"controlPlaneTaints":           clusterSpec.Cluster.Spec.ControlPlaneConfiguration.Taints,
 	}
+
+	if clusterSpec.Cluster.Spec.RegistryMirrorConfiguration != nil {
+		values["registryMirrorConfiguration"] = net.JoinHostPort(clusterSpec.Cluster.Spec.RegistryMirrorConfiguration.Endpoint, clusterSpec.Cluster.Spec.RegistryMirrorConfiguration.Port)
+		if len(clusterSpec.Cluster.Spec.RegistryMirrorConfiguration.CACertContent) > 0 {
+			values["registryCACert"] = clusterSpec.Cluster.Spec.RegistryMirrorConfiguration.CACertContent
+		}
+		// Added support for local registry, replace public.ecr.aws endpoint with the endpoint given in the cluster config file
```
Comment (chrisdoherty4) [inline pkg/providers/tinkerbell/template.go:418]:
```suggestion
		// Replace public.ecr.aws endpoint with the endpoint given in the cluster config file
```

File: test/framework/proxy.go [hunk 10]
```diff
@@ -14,6 +14,9 @@ const (
 	cloudstackHttpProxyVar  = "T_HTTP_PROXY_CLOUDSTACK"
 	cloudstackHttpsProxyVar = "T_HTTPS_PROXY_CLOUDSTACK"
 	cloudstackNoProxyVar    = "T_NO_PROXY_CLOUDSTACK"
+	TinkerbellHttpProxyVar  = "T_HTTP_PROXY_TINKERBELL"
```
Comment (chrisdoherty4) [inline test/framework/proxy.go:17]: Use upper-case for acronyms. `TinkerbellHTTPProxyVar`.

File: pkg/providers/tinkerbell/template.go [hunk 11]
```diff
@@ -402,10 +406,23 @@ func buildTemplateMapCP(clusterSpec *cluster.Spec, controlPlaneMachineSpec, etcd
 		"externalEtcdVersion":          bundle.KubeDistro.EtcdVersion,
 		"etcdCipherSuites":             crypto.SecureCipherSuitesString(),
 		"kubeletExtraArgs":             kubeletExtraArgs.ToPartialYaml(),
-		"controlPlanetemplateOverride": cpTemplateOverride,
 		"hardwareSelector":             controlPlaneMachineSpec.HardwareSelector,
 		"controlPlaneTaints":           clusterSpec.Cluster.Spec.ControlPlaneConfiguration.Taints,
 	}
+
+	if clusterSpec.Cluster.Spec.RegistryMirrorConfiguration != nil {
```
Comment (chrisdoherty4) [inline pkg/providers/tinkerbell/template.go:413]: Perhaps we could move this to its own function such as `populateRegistryMirrorValues(...)`.

File: pkg/providers/tinkerbell/stack/stack.go (hunk 8 above)
Comment (panktishah26) [reply pkg/providers/tinkerbell/stack/stack.go:108]: I needed pod cidr config and registry mirror config inside stack installer so I thought of passing a complete cluster configuration thinking we might need more info in future and adding a separate parameters for each values didn't sound right.

File: test/framework/proxy.go (hunk 10 above)
Comment (panktishah26) [reply test/framework/proxy.go:17]: Thanks for pointing out. I have removed e2e test related files from this PR and will create another PR for that.

File: pkg/providers/tinkerbell/stack/stack.go (hunk 8 above)
Comment (chrisdoherty4) [reply pkg/providers/tinkerbell/stack/stack.go:108]: I would give the installer what it needs and if we find we're continually adding stuff in the future we can adjust then.

File: pkg/dependencies/factory.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (panktishah26) [reply pkg/dependencies/factory.go]: @chrisdoherty4 I guess the above change would be covered in this ticket https://github.com/aws/eks-anywhere/issues/2977

[the remark this answers is not in the document]
Comment (chrisdoherty4) [reply pkg/dependencies/factory.go]: I don't think so. We're opting to specify `WithInsecure` here. If the code doesn't do what we need it to then we should address that.

That may result in the #2977 being being made redundant but I suspect not.

[the remark this answers is not in the document]
Comment (panktishah26) [reply pkg/dependencies/factory.go]: We are using insecure for helm chart at many places like tinkerbell-chart, cillium-chart, eks-anywhere-packages chart etc. I would like this to be addressed in a separate PR as it involves changes at many places.

[the remark this answers is not in the document]
Comment (ptrivedi) [reply pkg/dependencies/factory.go]: @chrisdoherty4 Not sure what you mean by #2977 becoming redundant with this. #2977 is meant to address this issue widely and generically

[the remark this answers is not in the document]
Comment (ptrivedi) [reply pkg/dependencies/factory.go]: I also think that we should let #2977 take care of the wider issue and not pollute or balloon the scope and focus of @panktishah26's PR

File: pkg/executables/config/kind.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (hitesh-infracloud) [reply pkg/executables/config/kind.yaml:43]: @panktishah26 

In which release this change will come into effect?

[the remark this answers is not in the document]
Comment (ptrivedi) [reply pkg/executables/config/kind.yaml:43]: > 
Hi Hitesh, Local Registry support for Bare Metal will be available in the 0.12 release of EKS Anywhere which is slated for later this month.
